File: Java/src/World/body/Animal.java
package World.body;

import GUI.Box;
import GUI.Color_obj;
import World.Point;
import World.World;

import java.io.Serializable;
import java.util.concurrent.ThreadLocalRandom;

public abstract class Animal extends Body implements Serializable {
    private static final long serialVersionUID = 738293893738294L;

    private Point last_position;
    private boolean random_location_born(Point new_point, Body other){
        int free_spaces = getWorld().free_spaces(this);
        int free_spaces_other = getWorld().free_spaces(other);
        if(free_spaces==0 && free_spaces_other==0){
            return false;
        }
        else if(free_spaces==0){
            new_point.setX(other.getPoint_location().getX());
            new_point.setY(other.getPoint_location().getY());
            while(true){
                random_location_empty(other, new_point);
                if(getWorld().getBoxes()[new_point.getY()-1][new_point.getX()-1].getColor().equals(Color_obj.EMPTY.getColor())){
                    return true;
                }
            }
        }
        else if(free_spaces_other==0){
            new_point.setX(getPoint_location().getX());
            new_point.setY(getPoint_location().getY());
            while(true){
                random_location_empty(this, new_point);
                if(getWorld().getBoxes()[new_point.getY()-1][new_point.getX()-1].getColor().equals(Color_obj.EMPTY.getColor())){
                    return true;
                }
            }
        }
        else{
            int randomNumber = ThreadLocalRandom.current().nextInt(1, 101) % 2;
            if(randomNumber==0){
                new_point.setX(getPoint_location().getX());
                new_point.setY(getPoint_location().getY());
                while(true){
                    random_location_empty(this, new_point);
                    if(getWorld().getBoxes()[new_point.getY()-1][new_point.getX()-1].getColor().equals(Color_obj.EMPTY.getColor())){
                        return true;
                    }
                }
            }
            else{
                new_point.setX(other.getPoint_location().getX());
                new_point.setY(other.getPoint_location().getY());
                while(true){
                    random_location_empty(other, new_point);
                    if(getWorld().getBoxes()[new_point.getY()-1][new_point.getX()-1].getColor().equals(Color_obj.EMPTY.getColor())){
                        return true;
                    }
                }
            }
        }
    }
    private void born(Body attacker){
        if(getAge()>3 && attacker.getAge()>3){
            Point new_point = new Point(getPoint_location().getX(), getPoint_location().getY());
            if(random_location_born(new_point, attacker)){
                new_body(new_point);
                getWorld().getNews_panel().add_news("Born new animal ("+new_point.getX()+","+new_point.getY()+")");
            }
            else{
                getWorld().getNews_panel().add_news("No place to born new animal");
            }
        }
        else{
            getWorld().getNews_panel().add_news("Animals are to young to born new animal");
        }
    }
    private void back_move(){
        int x = getPoint_location().getX();
        int y = getPoint_location().getY();
        setPoint_location(getLast_position());
        setLast_position(new Point(x,y));
        getWorld().getBoxes()[getPoint_location().getY()-1][getPoint_location().getX()-1].setColor(getColor());
    }
    protected boolean changed_position(){
        return !getLast_position().equals(getPoint_location());
    }

    public Animal() {
    }

    public Animal(Point point_location, World world) {
        super(point_location, world);
        setLast_position(point_location);
    }

    @Override
    public void move() {
        World world = getWorld();
        action();
        if(changed_position()){
            // jesli to pole jest puste
            if(!world.getBoxes()[getPoint_location().getY()-1][getPoint_location().getX()-1].getColor().equals(Color_obj.EMPTY.color)){
                //kolizja
                this.back_move();
                world.getBoxes()[getPoint_location().getY()-1][getPoint_location().getX()-1].setColor(Color_obj.EMPTY);
                Body tmp = world.get_body(getLast_position());
                if(tmp!=null){
                    this.back_move();
                    tmp.collision(this);
                }
            }
            else{
                world.getNews_panel().add_news("Move to ("+getPoint_location().getX()+","+getPoint_location().getY()+")");
                world.getBoxes()[getPoint_location().getY()-1][getPoint_location().getX()-1].setColor(getColor());
                world.getBoxes()[getLast_position().getY()-1][getLast_position().getX()-1].setColor(Color_obj.EMPTY);
            }
        }
        else{
            world.getNews_panel().add_news("Remains in place");
        }
    }

    @Override
    protected void collision(Body attacker) {
        if(attacker instanceof Animal){
            Animal other_animal = (Animal) attacker;
            if(other_animal.getColor().equals(getColor())){
                other_animal.back_move();;
                born(other_animal);
            }
            else{
                if(repel_attack(other_animal)){
                    getWorld().getNews_panel().add_news("Repel attack");
                }
                else{
                    if(getPower() > other_animal.getPower()){
                        getWorld().getNews_panel().add_news(this.getName()+" are killing "+other_animal.getName());
                        getWorld().delete_body2(other_animal);
                        getWorld().getBoxes()[getPoint_location().getY()-1][getPoint_location().getX()-1].setColor(this.getColor());
                    }
                    else{
                        getWorld().getNews_panel().add_news(other_animal.getName()+" are killing "+this.getName());
                        getWorld().delete_body(this);
                        getWorld().getBoxes()[other_animal.getPoint_location().getY()-1][other_animal.getPoint_location().getX()-1].setColor(other_animal.getColor());
                    }
                }
            }
        }
        else{
            if(getPower() > attacker.getPower()){
                getWorld().getNews_panel().add_news(attacker.getName()+" are eating "+this.getName()+" and die");
                getWorld().delete_body2(attacker);
                getWorld().getBoxes()[attacker.getPoint_location().getY()-1][attacker.getPoint_location().getX()-1].setColor(this.getColor());
            }
            else{
                getWorld().getNews_panel().add_news(this.getName()+" are eating "+attacker.getName());
                getWorld().delete_body(this);
                getWorld().getBoxes()[getPoint_location().getY()-1][getPoint_location().getX()-1].setColor(attacker.getColor());
            }
        }
    }

    @Override
    protected void action() {
        int x = getPoint_location().getX();
        int y = getPoint_location().getY();
        Point point = new Point(x ,y);
        random_location(point);
        setLast_position(getPoint_location());
        setPoint_location(point);
    }

    protected abstract boolean repel_attack(Body attacker);

    public void setLast_position(Point last_position) {
        this.last_position = last_position;
    }

    public Point getLast_position() {
        return last_position;
    }
}

File: Python/World/Body/animal.py
from abc import abstractmethod, ABC
from random import randint

from GUI.colors import Colors
from World.Body.body import Body
from World.point import Point


class Animal(Body, ABC):
    last_position = None

    def __init__(self, point_location, world):
        super().__init__(point_location, world)
        self.set_last_position(self.get_point_location())

    def _random_location_born(self, new_point, other):
        free_spaces = self.get_world().free_spaces(self)
        free_spaces_other = self.get_world().free_spaces(other)
        if free_spaces == 0 and free_spaces_other == 0:
            return False
        elif free_spaces == 0:
            new_point.x = other.get_point_location().x
            new_point.y = other.get_point_location().y
            while True:
                self.random_location_empty(other, new_point)
                if self.get_world().get_boxes[new_point.y - 1][new_point.x - 1].get_color() == Colors.EMPTY:
                    return True
        elif free_spaces_other == 0:
            new_point.x = self.get_point_location().x
            new_point.y = self.get_point_location().y
            while True:
                self.random_location_empty(self, new_point)
                if self.get_world().get_boxes[new_point.y - 1][new_point.x - 1].get_color() == Colors.EMPTY:
                    return True
        else:
            randomNumber = randint(1, 101) % 2
            if randomNumber > 50:
                new_point.x = self.get_point_location().x
                new_point.y = self.get_point_location().y
                while True:
                    self.random_location_empty(self, new_point)
                    if self.get_world().get_boxes[new_point.y - 1][new_point.x - 1].get_color() == Colors.EMPTY:
                        return True
            else:
                new_point.x = other.get_point_location().x
                new_point.y = other.get_point_location().y
                while True:
                    self.random_location_empty(other, new_point)
                    if self.get_world().get_boxes[new_point.y - 1][new_point.x - 1].get_color() == Colors.EMPTY:
                        return True

    def _born(self, attacker):
        if self.get_age() > 3 and attacker.get_age() > 3:
            new_point = Point(self.get_point_location().get_x(), self.get_point_location().get_y())
            if self._random_location_born(new_point, attacker):
                self.new_body(new_point)
                # self.get_world().getNews_panel().add_news(
                # "Born new animal (" + str(new_point.get_x()) + "," + str(new_point.get_y()) + ")")
        #     else:
        #         # self.get_world().getNews_panel().add_news("No place to born new animal")
        # else:
        #     # self.get_world().getNews_panel().add_news("Animals are too young to born new animal")

    def _back_move(self):
        x = self.point_location.get_x()
        y = self.point_location.get_y()
        self.set_point_location(self.get_last_position())
        self.set_last_position(Point(x, y))
        self.world.get_boxes[self.point_location.y - 1][self.point_location.x - 1].set_color(self.color)

    def action(self):
        x = self.point_location.x
        y = self.point_location.y
        point = Point(x, y)
        self.random_location(point)
        self.set_last_position(self.get_point_location())
        self.set_point_location(point)

    def move(self):
        world = self.get_world()
        self.action()
        if self.changed_position():
            if world.get_boxes[self.get_point_location().get_y() - 1][self.get_point_location().get_x() - 1].get_color() != Colors.EMPTY:
                self._back_move()
                world.get_boxes[self.point_location.y - 1][self.point_location.x - 1].set_color(Colors.EMPTY)
                tmp = world.get_body(self.last_position)
                if tmp is not None:
                    self._back_move()
                    tmp.collision(self)
            else:
                # world.get_news_panel().add_news("Move to ({}, {})".format(self.point_location.x, self.point_location.y))
                world.get_boxes[self.point_location.y - 1][self.point_location.x - 1].set_color(self.get_color())
                world.get_boxes[self.last_position.y - 1][self.last_position.x - 1].set_color(Colors.EMPTY)

    def collision(self, attacker):
        if isinstance(attacker, Animal):
            other_animal = attacker
            if other_animal.get_color() == self.get_color():
                other_animal._back_move()
                self._born(other_animal)
            else:
                if self.repel_attack(other_animal):
                    print("repel attack")
                    # self.world.get_news_panel().add_news("Repel attack")
                else:
                    if self.get_power() > other_animal.get_power():
                        # self.world.get_news_panel().add_news(self.get_name() + " is killing " + other_animal.get_name())
                        self.world.delete_body2(other_animal)
                        self.world.get_boxes[self.point_location.y - 1][
                            self.point_location.x - 1].set_color(self.get_color())
                    else:
                        # self.world.get_news_panel().add_news(other_animal.get_name() + " is killing " + self.get_name())
                        self.world.delete_body(self)
                        self.world.get_boxes[other_animal.get_point_location().y - 1][
                            other_animal.get_point_location().x - 1].set_color(other_animal.get_color())
        else:
            if self.get_power() > attacker.get_power():
                # self.world.get_news_panel().add_news(
                    # attacker.get_name() + " is eating " + self.get_name() + " and dies")
                self.world.delete_body2(attacker)
                self.world.get_boxes[attacker.get_point_location().y - 1][
                    attacker.get_point_location().x - 1].set_color(self.get_color())
            else:
                # self.world.get_news_panel().add_news(self.get_name() + " is eating " + attacker.get_name())
                self.world.delete_body(self)
                self.world.get_boxes[self.point_location.y - 1][
                    self.point_location.x - 1].set_color(attacker.get_color())

    def changed_position(self):
        return self.last_position != self.point_location

    def set_last_position(self, last_position):
        self.last_position = last_position

    def get_last_position(self):
        return self.last_position

    @abstractmethod
    def repel_attack(self, attacker):
        pass
